hadoop-tools/hadoop-azure/src/test/java/org/apache/hadoop/fs/azurebfs/services/TestAbfsRestOperation.java

```diff
@@ -0,0 +1,360 @@
+    Mockito.doReturn(HTTP_METHOD_PUT).when(abfsHttpOperation).getConnRequestMethod();
+
+    switch (errorType) {
+    case OUTPUTSTREAM:
+      // If the getOutputStream() throws IOException and Expect Header is
+      // enabled, it returns back to processResponse and hence we have
+      // mocked the response code and the response message to check different
+      // behaviour based on response code.
+      Mockito.doReturn(responseCode).when(abfsHttpOperation).getConnResponseCode();
+      Mockito.doReturn(responseMessage)
+          .when(abfsHttpOperation)
+          .getConnResponseMessage();
+      Mockito.doThrow(new ProtocolException("Server rejected Operation"))
+          .when(abfsHttpOperation)
+          .getConnOutputStream();
+      break;
+    case WRITE:
+      // If write() throws IOException and Expect Header is
+      // enabled or not, it should throw back the exception.
+      OutputStream outputStream = Mockito.spy(new OutputStream() {
+        @Override
+        public void write(final int i) throws IOException {
+        }
+      });
+      Mockito.doReturn(outputStream).when(abfsHttpOperation).getConnOutputStream();
+      Mockito.doThrow(new IOException())
+          .when(outputStream)
+          .write(buffer, appendRequestParameters.getoffset(),
+              appendRequestParameters.getLength());
+      break;
+    default:
+      break;
+    }
+
+    // Sets the httpOperation for the rest operation.
+    Mockito.doReturn(abfsHttpOperation)
+        .when(op)
+        .getHttpOperation(Mockito.any(), Mockito.any(), Mockito.any());
+    return op;
+  }
+
+  /**
+   * Test the functionalities based on whether getOutputStream() or write()
+   * throws exception and what is the corresponding response code.
+   */
+  @Test
+  public void testExpectHundredContinue() throws Exception {
+    // Gets the AbfsRestOperation.
+    AbfsRestOperation op = getRestOperation();
+    AbfsHttpOperation httpOperation = op.getHttpOperation(Mockito.any(), Mockito.any(), Mockito.any());
+
+    TracingContext tracingContext = Mockito.spy(new TracingContext("abcd",
+        "abcde", FSOperationType.APPEND,
+        TracingHeaderFormat.ALL_ID_FORMAT, null));
+
+    switch (errorType) {
+    case WRITE:
+      // If write() throws IOException and Expect Header is
+      // enabled or not, it should throw back the exception
+      // which is caught and exponential retry logic comes into place.
+      intercept(IOException.class,
+          () -> op.execute(tracingContext));
+
+      // Assert that the request is retried based on reduced retry count configured.
+      Assertions.assertThat(tracingContext.getRetryCount())
+          .describedAs("The retry count is incorrect")
+          .isEqualTo(REDUCED_RETRY_COUNT);
+
+      // Assert that metrics will be updated correctly.
+      Assertions.assertThat(httpOperation.getBytesSent())
+          .isEqualTo(BUFFER_LENGTH);
+      break;
+    case OUTPUTSTREAM:
```

Comment:
case in case is pretty complex; in production I'd be worried. tests, well, it works.